pkg/cached/cache.go

```diff
@@ -0,0 +1,257 @@
+// merging or transforming function.
+type Data[T any] interface {
+	// Returns the cached data, as well as an "etag" to identify the
+	// version of the cache, or an error if something happened.
+	Get() Result[T]
+}
+
+// T is the source type, V is the destination type.
+type merger[K comparable, T, V any] struct {
+	mergeFn      func(map[K]Result[T]) Result[V]
+	caches       map[K]Data[T]
+	cacheResults map[K]Result[T]
+	result       Result[V]
+}
+
+// NewMerger creates a new merge cache, a cache that merges the result
+// of other caches. The function only gets called if any of the
+// dependency has changed.
+//
+// If any of the dependency returned an error before, or any of the
+// dependency returned an error this time, or if the mergeFn failed
+// before, then the function is reran.
+//
+// The caches and results are mapped by string so that associated data
+// can be retrieved.
+func NewMerger[K comparable, T, V any](mergeFn func(results map[K]Result[T]) Result[V], caches map[K]Data[T]) Data[V] {
+	return &merger[K, T, V]{
+		mergeFn: mergeFn,
+		caches:  caches,
+	}
+}
+
+func (c *merger[K, T, V]) prepareResults() map[K]Result[T] {
+	cacheResults := make(map[K]Result[T], len(c.caches))
+	for key, cache := range c.caches {
+		cacheResults[key] = cache.Get()
+	}
+	return cacheResults
+}
+
+// Rerun if:
+//   - The last run resulted in an error
+//   - Any of the dependency previously returned an error
+//   - Any of the dependency just returned an error
+//   - Any of the dependency's etag changed
+func (c *merger[K, T, V]) needsRunning(results map[K]Result[T]) bool {
+	if c.cacheResults == nil {
+		return true
+	}
+	if c.result.Err != nil {
+		return true
+	}
+	if len(results) != len(c.cacheResults) {
+		panic(fmt.Errorf("invalid number of results: %v (expected %v)", len(results), len(c.cacheResults)))
+	}
+	for key, oldResult := range c.cacheResults {
+		newResult, ok := results[key]
+		if !ok {
+			panic(fmt.Errorf("unknown cache entry: %v", key))
+		}
+
+		if newResult.Etag != oldResult.Etag || newResult.Err != nil || oldResult.Err != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *merger[K, T, V]) Get() Result[V] {
+	cacheResults := c.prepareResults()
+	if c.needsRunning(cacheResults) {
+		c.cacheResults = cacheResults
+		c.result = c.mergeFn(c.cacheResults)
+	}
+	return c.result
+}
+
+const transformerCacheName = "transformer"
+
+// NewTransformer creates a new cache that transforms the result of
+// another cache. The transformFn will only be called if the source
+// cache has updated the output, otherwise, the cached result will be
+// returned.
+//
+// If the dependency returned an error before, or it returns an error
+// this time, or if the transformerFn failed before, the function is
+// reran.
+func NewTransformer[T, V any](transformerFn func(Result[T]) Result[V], source Data[T]) Data[V] {
+	return NewMerger(func(caches map[string]Result[T]) Result[V] {
```

Comment:
instead of string, you could make a `type transformerCacheKeyType struct{}` and be guaranteed that there is at most one entry in the map :)